File: src/output.rs
use chrono::{DateTime, Utc};
use comfy_table::{modifiers::UTF8_ROUND_CORNERS, presets::UTF8_FULL, Table};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum OutputMode {
    Human,
    Json,
}

impl OutputMode {
    pub fn from_flag(json: bool) -> Self {
        if json {
            Self::Json
        } else {
            Self::Human
        }
    }
}

/// Print a value as JSON or as a formatted table.
pub fn print_output<T: serde::Serialize>(
    mode: OutputMode,
    value: &T,
    table_fn: impl FnOnce(&T) -> Table,
) {
    match mode {
        OutputMode::Json => {
            let json = serde_json::to_string_pretty(value).expect("JSON serialization failed");
            println!("{json}");
        }
        OutputMode::Human => {
            let table = table_fn(value);
            println!("{table}");
        }
    }
}

/// Print a list of values as JSON array or as a table with rows.
pub fn print_list<T: serde::Serialize>(
    mode: OutputMode,
    items: &[T],
    headers: &[&str],
    row_fn: impl Fn(&T) -> Vec<String>,
) {
    match mode {
        OutputMode::Json => {
            let json = serde_json::to_string_pretty(items).expect("JSON serialization failed");
            println!("{json}");
        }
        OutputMode::Human => {
            if items.is_empty() {
                println!("No results.");
                return;
            }
            let mut table = new_table();
            table.set_header(headers.iter().copied());
            for item in items {
                table.add_row(row_fn(item));
            }
            println!("{table}");
        }
    }
}

pub fn new_table() -> Table {
    let mut table = Table::new();
    table
        .load_preset(UTF8_FULL)
        .apply_modifier(UTF8_ROUND_CORNERS);
    table
}

pub fn format_time(dt: &DateTime<Utc>) -> String {
    dt.format("%Y-%m-%d %H:%M:%S").to_string()
}

pub fn format_duration_secs(secs: Option<f64>) -> String {
    match secs {
        Some(s) if s < 60.0 => format!("{s:.1}s"),
        Some(s) => {
            let mins = (s / 60.0).floor();
            let remainder = s - mins * 60.0;
            format!("{mins:.0}m {remainder:.1}s")
        }
        None => "-".to_string(),
    }
}

pub fn short_uuid(id: &uuid::Uuid) -> String {
    let s = id.to_string();
    s[..8].to_string()
}

pub fn format_bytes(bytes: i64) -> String {
    const KB: f64 = 1024.0;
    const MB: f64 = KB * 1024.0;
    const GB: f64 = MB * 1024.0;
    #[expect(
        clippy::cast_precision_loss,
        reason = "file sizes don't need i64 precision"
    )]
    let b = bytes as f64;
    if b < KB {
        format!("{bytes} B")
    } else if b < MB {
        format!("{:.1} KB", b / KB)
    } else if b < GB {
        format!("{:.1} MB", b / MB)
    } else {
        format!("{:.2} GB", b / GB)
    }
}

File: lib/output.js
// Copyright 2026 Hemi Labs, Inc.
// SPDX-License-Identifier: GPL-3.0-only

import Table from "cli-table3";

export function printOutput(json, value, tableFn) {
  if (json) {
    console.log(JSON.stringify(value, null, 2));
  } else {
    const table = tableFn(value);
    console.log(table.toString());
  }
}

export function printList(json, items, headers, rowFn) {
  if (json) {
    console.log(JSON.stringify(items, null, 2));
    return;
  }
  if (items.length === 0) {
    console.log("No results.");
    return;
  }
  const table = newTable();
  table.push(headers);
  for (const item of items) {
    table.push(rowFn(item));
  }
  console.log(table.toString());
}

export function newTable() {
  return new Table({
    chars: {
      top: "\u2500", "top-mid": "\u252c", "top-left": "\u256d", "top-right": "\u256e",
      bottom: "\u2500", "bottom-mid": "\u2534", "bottom-left": "\u2570", "bottom-right": "\u256f",
      left: "\u2502", "left-mid": "\u251c", mid: "\u2500", "mid-mid": "\u253c",
      right: "\u2502", "right-mid": "\u2524", middle: "\u2502",
    },
  });
}

export function kvTable(rows) {
  const table = newTable();
  table.push(["FIELD", "VALUE"]);
  for (const [k, v] of rows) {
    table.push([k, v]);
  }
  return table;
}

export function formatTime(dt) {
  if (!dt) {
    return "-";
  }
  const d = new Date(dt);
  const pad = (n) => String(n).padStart(2, "0");
  return `${d.getUTCFullYear()}-${pad(d.getUTCMonth() + 1)}-${pad(d.getUTCDate())} ${pad(d.getUTCHours())}:${pad(d.getUTCMinutes())}:${pad(d.getUTCSeconds())}`;
}

export function formatDurationSecs(secs) {
  if (secs === null || secs === undefined) {
    return "-";
  }
  if (secs < 60) {
    return `${secs.toFixed(1)}s`;
  }
  const mins = Math.floor(secs / 60);
  const remainder = secs - mins * 60;
  return `${mins}m ${remainder.toFixed(1)}s`;
}

export function formatBytes(bytes) {
  const KB = 1024;
  const MB = KB * 1024;
  const GB = MB * 1024;
  if (bytes < KB) {
    return `${bytes} B`;
  }
  if (bytes < MB) {
    return `${(bytes / KB).toFixed(1)} KB`;
  }
  if (bytes < GB) {
    return `${(bytes / MB).toFixed(1)} MB`;
  }
  return `${(bytes / GB).toFixed(2)} GB`;
}

export function formatRuntime(secs) {
  const hours = Math.floor(secs / 3600);
  const mins = Math.floor((secs % 3600) / 60);
  if (mins > 0) {
    return `${hours}h${mins}m`;
  }
  return `${hours}h`;
}

export function formatMemory(mb) {
  if (mb >= 1024 && mb % 1024 === 0) {
    return `${mb / 1024} GB`;
  }
  return `${mb} MB`;
}
